packages/notebook/src/history.ts

```diff
@@ -0,0 +1,420 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { Cell } from '@jupyterlab/cells';
+import { ISessionContext } from '@jupyterlab/apputils';
+import { CodeEditor } from '@jupyterlab/codeeditor';
+import { KernelMessage } from '@jupyterlab/services';
+import { ITranslator, nullTranslator } from '@jupyterlab/translation';
+import { IDisposable } from '@lumino/disposable';
+import { Signal } from '@lumino/signaling';
+import { IKernelConnection } from '@jupyterlab/services/lib/kernel/kernel';
+
+/**
+ * The definition of a console history manager object.
+ */
+export interface INotebookHistory extends IDisposable {
+  /**
+   * The session context used by the foreign handler.
+   */
+  readonly sessionContext: ISessionContext;
+
+  /**
+   * Translator to be used for warnings.
+   */
+  readonly translator: ITranslator;
+
+  /**
+   * The current editor used by the history widget.
+   */
+  editor: CodeEditor.IEditor | null;
+
+  /**
+   * The placeholder text that a history session began with.
+   */
+  readonly placeholder: string;
+
+  /**
+   * The session number of the current kernel session
+   */
+  readonly kernelSession: string;
+
+  /**
+   * Get the previous item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  back(activeCell: Cell): Promise<string>;
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  forward(activeCell: Cell): Promise<string>;
+
+  /**
+   * Reset the history navigation state, i.e., start a new history session.
+   */
+  reset(): void;
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the result from back or forward
+   */
+  updateEditor(activeCell: Cell, update: string): void;
+}
+
+/**
+ * A console history manager object.
+ */
+export class NotebookHistory implements INotebookHistory {
+  /**
+   * Construct a new console history object.
+   */
+  constructor(options: NotebookHistory.IOptions) {
+    this.sessionContext = options.sessionContext;
+    this.translator = options.translator || nullTranslator;
+    void this._handleKernel().then(() => {
+      this.sessionContext.kernelChanged.connect(this._handleKernel, this);
+    });
+    this.toRequest = this.requestBatchSize;
+  }
+
+  /**
+   * The client session used by the foreign handler.
+   */
+  readonly sessionContext: ISessionContext;
+
+  /**
+   * Translator to be used for warnings
+   */
+  readonly translator: ITranslator;
+
+  /**
+   * The number of history items to request.
+   */
+  private toRequest: number;
```

Comment:
Nitpick: elsewhere we use underscore prefix for private properties (though not strictly required).